### PR TITLE
Add rustfmt

### PR DESCRIPTION
**Pull request by iwburns**

### CODE REVIEW

**Comment by iwburns** [comment]
@Drakulix I've finally gotten around to setting this up.

Any other options you can think of that I might have left out?  I know this is mostly personal preference, but I'd still like to hear your opinion on these if you feel like it.

There's one bit of formatting that I'm not sure how I feel about, but I'll add an in-line comment for it shortly instead of trying to describe it here.

File: src/tree.rs [hunk 1]
```diff
@@ -714,7 +717,8 @@ impl<T> Tree<T> {
     /// # }
     /// ```
     ///
-    pub fn sort_children_by_key<B, F>(&mut self, node_id: &NodeId, mut f: F) -> Result<(), NodeIdError>
+    pub fn sort_children_by_key<B, F>(&mut self, node_id: &NodeId, mut f: F)
+                                      -> Result<(), NodeIdError>
         where B: Ord, F: FnMut(&Node<T>) -> B
     {
```
Comment (iwburns) [inline src/tree.rs]: I'm not sure how I feel about the formatting when we have a `return` that gets pushed to the next line and we also have a where clause (which is always pushed to the next line).

I feel like I would prefer it to be something like:
```
pub fn sort_children_by_key<B, F>(&mut self, node_id: &NodeId, mut f: F)
    -> Result<(), NodeIdError>
    where B: Ord, F: FnMut(&Node<T>) -> B
{
```
in that case, but I'm not really familiar enough with rustfmt to know what combinations of settings would produce that (if it's even possible) while leaving the rest of the behavior alone.

Any thoughts on this @Drakulix ?

Comment (iwburns) [inline src/tree.rs]: Similarly, in a situation where there's just a `return` value that gets pushed to the next line but there's no `where` clause,  I think it would probably be better to have something like:
```
pub fn sort_children_by_key<B, F>(&mut self, node_id: &NodeId, mut f: F)
    -> Result<(), NodeIdError>
{
```
instead of having the `{` on the same line as the `return` value and having it tabbed over to line up with the function args.

Comment (Drakulix) [inline src/tree.rs]: I am not sure, if I know rustfmt good enough for this specific. Of course it has to do with the maximum line length, but in case you want to keep that, I am not sure, if you can control the return value indentation.
You could probably fill an issue on rustfmt. If there is an option and we cannot find it, this is at least a documentation/discoverability problem.

Comment (iwburns) [inline src/tree.rs]: Yeah, I actually did end up submitting an [issue](https://github.com/rust-lang-nursery/rustfmt/issues/1260), so we'll just see if that goes anywhere.

File: src/tree.rs [hunk 2]
```diff
@@ -228,8 +227,11 @@ impl<T> Tree<T> {
     /// tree.insert_with_parent(child_node, &root_id);
     /// ```
     ///
+    #[cfg_attr(rustfmt, rustfmt_skip)]
+    // todo: remove this if https://github.com/rust-lang-nursery/rustfmt/issues/1260 is resolved
     pub fn insert_with_parent(&mut self, child: Node<T>, parent_id: &NodeId)
-        -> Result<NodeId, NodeIdError> {
+        -> Result<NodeId, NodeIdError>
+    {
```
Comment (iwburns) [inline src/tree.rs]: @Drakulix this is what I went with for now.  Seem alright to you?

Comment (Drakulix) [inline src/tree.rs]: It could look way worse. This is actually quite nice.

**Comment by iwburns** [comment]
The only thing I'm wondering about now on this one is whether or not to have Travis fail if there are formatting issues.

Part of me thinks doing that might get a bit ridiculous for certain things.  For example, this would probably fail right now due to some `error.expect("[some long and very specific error message]")` lines.  I don't really want to change those because they might come in handy at some point, but I also don't really like the idea of having to include `#[cfg_attr(rustfmt, rustfmt_skip)]` right before each of those lines.

Basically, I'm okay with ignoring those specific failures, but otherwise, I'd like Travis to report it.  I guess the only way to do that is to add the `rustfmt_skip`, I just don't like that solution very much.

Thoughts @Drakulix ?

**Comment by Drakulix** [comment]
I dont have a better solution. I think you need to go with `rustfmt_skip` in that case

**Comment by iwburns** [comment]
Well, it seems like using `rustfmt_skip` won't quite get me what I want, so I just decided to use multi-line strings instead.

**Comment by Drakulix** [comment]
Well it doesn't look to bad either. Thanks for addressing this!

**Comment by iwburns** [comment]
Yeah no problem!  And to be honest, I forgot that multi-line strings worked that way in Rust (I almost never use them), but like you said, it doesn't look too bad.
